### PR TITLE
SIMD enablement

### PR DESCRIPTION
**Pull request by shawnl**
There are some rough parts here:

 * should shuffle a and b arguments implicitly cast from arrays? (I think not, as it wouldn't be useful, and it would have higher alignment requirements than regular arrays).
 * the `all()` and `any()` builtin functions on vectors of bools does not take parentheses. (We can open a tracking bug for this).
 * no compiler error tests for the out-of-bounds vector access.

But what is here works, and perhaps should be merged to remove merge pain. I plan on doing much more work, and getting some MVP/POC stuff out soon, to show how useful this stuff is.

Again, review one patch at a time. The github review interface does not really support that, which is a shame.

### CODE REVIEW

File: doc/langref.html.in [hunk 1]
```diff
@@ -7952,6 +7952,25 @@ pub const TypeInfo = union(TypeId) {
       {#link|pointer|Pointers#}.
       </p>
       {#header_close#}
+
+      {#header_open|@shuffle#}
+      <pre>{#syntax#}@shuffle(comptime ElemType: type, a: @Vector(mask.len, ElemType), b: @Vector(mask.len, ElemType), comptime mask: @Vector(_, u32)) @Vector(mask.len, ElemType){#endsyntax#}</pre>
+      <p>
+      Does the {#syntax#}shufflevector{#endsyntax#} instruction. Each element in {#syntax#}comptime{#endsyntax#}
```
Comment (daurnimator) [inline doc/langref.html.in]: The `shufflevector` instruction is not standard or widely known, so it should be defined here.

File: test/stage1/behavior/vector.zig [hunk 2]
```diff
@@ -74,3 +74,67 @@ test "implicit cast vector to array" {
...
+            expect(((v  > x) == @Vector(4, bool)([4]bool{  true, false, false,  true})).all);
+            expect(((v <= x) == @Vector(4, bool)([4]bool{ false,  true,  true, false})).all);
+            expect(((v >= x) == @Vector(4, bool)([4]bool{  true, false,  true,  true})).all);
+            expect(@Vector(4, bool)([4]bool{ false, false, false, false}).any == false);
+            expect(@Vector(4, bool)([4]bool{ true, false, false, false}).any);
+        }
+    };
+    S.doTheTest();
+    comptime S.doTheTest();
+}
```
Comment (daurnimator) [inline test/stage1/behavior/vector.zig]: missing newline at end of file

File: test/stage1/behavior/shuffle.zig [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            // Upcasting of a
+            var v3: @Vector(2, i32) = [2]i32{ 2147483647, -2};
+            const mask4: @Vector(4, i32) = [4]i32{ 0, ~i32(2), 1, ~i32(3)};
+            res = @shuffle(i32, v3, x, mask4);
+            expect(mem.eql(i32, ([4]i32)(res), [4]i32{ 2147483647, 3, -2, 4 }));
+        }
+    };
+    S.doTheTest();
+    comptime S.doTheTest();
+}
```
Comment (daurnimator) [inline test/stage1/behavior/shuffle.zig]: missing newline

File: src/codegen.cpp [hunk 4]
```diff
@@ -3422,6 +3440,28 @@ static LLVMValueRef ir_render_load_ptr(CodeGen *g, IrExecutable *executable, IrI
...
+        LLVMValueRef i_val = LLVMConstInt(g->builtin_types.entry_i32->llvm_type, i, false);
+        LLVMValueRef extract = LLVMBuildExtractElement(g->builder, instruction->vector->llvm_value, i_val, "");
+        if (instruction->is_any)
+            result = LLVMBuildOr(g->builder, result, extract, "test_any");
+        else
+            result = LLVMBuildAnd(g->builder, result, extract, "test_all");
+    }
+    LLVMValueRef zero = LLVMConstNull(LLVMTypeOf(result));
+    result = LLVMBuildICmp(g->builder, LLVMIntUGT, result, zero, "");
+    // I am not sure that this alloca should be necessary, but currently it is
```
Comment (daurnimator) [inline src/codegen.cpp]: This is missing copy elison?

File: src/codegen.cpp [hunk 5]
```diff
@@ -4226,6 +4277,35 @@ static LLVMValueRef ir_render_ctz(CodeGen *g, IrExecutable *executable, IrInstru
...
 }
 
+static LLVMValueRef ir_render_shuffle_vector(CodeGen *g, IrExecutable *executable, IrInstructionShuffleVector *instruction) {
+    uint64_t len_a = instruction->a->value.type->data.vector.len,
+        len_c = instruction->mask->value.type->data.vector.len;
+
+    // LLVM uses integers larger than the length of the first array to
+    // index into the second array. This was deemed unnecessarily fragile
+    // when changing code, so Zig uses negative numbers to index the
+    // second vector. These start at -1 and go down, are are easiest to use
```
Comment (euantorano) [inline src/codegen.cpp]: `are are` should be `and are` I believe?

```suggestion
    // second vector. These start at -1 and go down, and are easiest to use
```

**Comment by shawnl** [comment]
Is there a way for me to get access to a OSX or Windows VPS to debug this? Ever better, without having to set up a Zig build environment.

File: doc/langref.html.in (hunk 1 above)
Comment (shawnl) [reply doc/langref.html.in]: It is defined in the next sentence....

File: src/codegen.cpp (hunk 4 above)
Comment (shawnl) [reply src/codegen.cpp]: ugggh, if we just passed by value then it wouldn't be so complicated.

File: test/stage1/behavior/vector.zig (hunk 2 above)
Comment (shawnl) [reply test/stage1/behavior/vector.zig]: Not only are complaints like this annoying (just fix zig fmt and don't do distracting style reviews), but git actually warns if you have a blank line at EOF

`<stdin>:12: new blank line at EOF.`

Comment (shawnl) [reply test/stage1/behavior/vector.zig]: And the only reason I was going to fix this is to make you stop correcting me

File: src/codegen.cpp (hunk 4 above)
Comment (shawnl) [reply src/codegen.cpp]: IrInstructionBinOp does not have a `ResultLoc`, so I'm not sure which pattern I should follow. @andrewrk  ?

File: test/stage1/behavior/vector.zig (hunk 2 above)
Comment (daurnimator) [reply test/stage1/behavior/vector.zig]: IMO, until we run `zig fmt` as part of CI or merging, it needs to be caught on review.

> but git actually warns if you have a blank line at EOF

That would be a *blank* line at end of file; this file was missing the terminating newline.

Comment (shawnl) [reply test/stage1/behavior/vector.zig]: > That would be a blank line at end of file; this file was missing the terminating newline. (which git should warn you about missing!)

Nope. Maybe you should try it.

Comment (shawnl) [reply test/stage1/behavior/vector.zig]: Oh I see. I'll report a bug will vscode.

**Comment by shawnl** [comment]
This is now a draft PR, but it doesn't seem like I can convert it.

**Comment by shawnl** [comment]
This still requires some work, and I cannot convert this PR to a draft PR.

**Pull request closed without merging**